src/ng_can.c

```diff
@@ -69,61 +69,108 @@ static struct canfd_frame parse_can_frame(const char *req, int *req_index)
       errx(EXIT_FAILURE, "Bad Can ID");
 
     long data_len;
-    if(ei_decode_binary(req, req_index, can_frame.data, &data_len) < 0 || data_len > CANFD_MAX_DLEN)
+    if(ei_decode_binary(req, req_index, can_frame.data, &data_len) < 0 || data_len > CAN_MAX_DLEN)
       errx(EXIT_FAILURE, "Bad Data");
 
     can_frame.can_id = id;
-    can_frame.len = data_len;
+    can_frame.can_dlc = data_len;
 
     return can_frame;
 }
 
+static struct canfd_frame parse_canfd_frame(const char *req, int *req_index)
+{
+    struct canfd_frame canfd_frame = { 0 };
+
+    int num_tuple_elements;
+    if(ei_decode_tuple_header(req, req_index, &num_tuple_elements) < 0 || num_tuple_elements != 2)
+      errx(EXIT_FAILURE, "Bad Tuple");
+
+    unsigned long id;
+    if(ei_decode_ulong(req, req_index, &id) < 0)
+      errx(EXIT_FAILURE, "Bad Can ID");
+
+    long data_len;
+    if(ei_decode_binary(req, req_index, canfd_frame.data, &data_len) < 0 || data_len > CANFD_MAX_DLEN)
+      errx(EXIT_FAILURE, "Bad Data");
+
+    canfd_frame.can_id = id;
+    canfd_frame.len = data_len;
+
+    return canfd_frame;
+}
+
 static int write_buffer(const char *req, int *req_index, int num_frames)
 {
-  for (int i = 0; i < num_frames; i++) {
+  bool is_canfd = can_port->is_canfd;
+
+  for (int i = 0; i < num_frames; i++)
+  {
     int this_frame_offset = *req_index;
-    struct canfd_frame can_frame = parse_can_frame(req, req_index);
-    int write_result = can_write(can_port, &can_frame);
+    int write_result;
 
-    //ENETDOWN is okay since we're restarting using `ip link` in ng_can.ex?
-    if(write_result < 0 && (errno == EAGAIN || errno == ENOBUFS || errno == ENETDOWN)) {
+    if (is_canfd)
+    {
+      struct canfd_frame canfd_frame = parse_canfd_frame(req, req_index);
+      write_result = canfd_write(can_port, &canfd_frame);
+    }
+    else
+    {
+      struct can_frame can_frame = parse_can_frame(req, req_index);
+      write_result = can_write(can_port, &can_frame);
+    }
+    
+    if(write_result < 0 && (errno == EAGAIN || errno == ENOBUFS))
```

Comment:
I removed the ENETDOWN case as I also removed the elixir code trying to set up the interfaces.